fdbclient/include/fdbclient/DatabaseContext.h

```diff
@@ -685,8 +705,16 @@ class DatabaseContext : public ReferenceCounted<DatabaseContext>, public FastAll
 	EventCacheHolder connectToDatabaseEventCacheHolder;
 
 private:
-	std::unordered_map<std::pair<int64_t, Key>, Reference<WatchMetadata>, boost::hash<std::pair<int64_t, Key>>>
-	    watchMap;
+	using WatchMapKey = std::pair<int64_t, Key>;
+	using WatchMapKeyHasher = boost::hash<WatchMapKey>;
+	using WatchMapValue = Reference<WatchMetadata>;
+	using WatchMap_t = std::unordered_map<WatchMapKey, WatchMapValue, WatchMapKeyHasher>;
+
+	WatchMap_t watchMap;
+
+	using WatchCounterMap_t = std::unordered_map<WatchMapKey, int32_t, WatchMapKeyHasher>;
+	// Maps the number of the WatchMapKey being used.
+	WatchCounterMap_t watchCounterMap;
```

Comment:
Instead of keeping a separate map, could we add a counter field to `WatchMetadata`?

Reply:
This would be hard as `WatchMetadata` could be reconstructed in certain conditions, e.g. a new watch on the same key/value pair or different value, etc. Storing it inside the `WatchMetadata` requires tracing all such reconstructions. For this reason, I tend to put it in parallel to the `map` in the `DatabaseContext`. I do not have a strong opinion on the issue if I should use a `pair` or create a new `unordered_map`, though; here I used a new `unordered_map`.

Reply:
Looking at the cases where we replace the existing metadata, it seems like this should be fairly trivial. If I understand correctly we are either creating a new metadata or the old one is being triggered, so we don't need to maintain the old reference count.